File: frontend/src/engine/math/angles.ts
/** Angle helpers. Serialized angles are counterclockwise degrees. */

export function degToRad(degrees: number): number {
  return (degrees * Math.PI) / 180;
}

export function radToDeg(radians: number): number {
  return (radians * 180) / Math.PI;
}

/**
 * Normalize to the half-open interval (-180, 180].
 * JavaScript `%` truncates toward zero exactly like Python `math.fmod`,
 * so both kernels wrap identically.
 */
export function normalizeDeg(degrees: number): number {
  let wrapped = degrees % 360;
  if (wrapped > 180) {
    wrapped -= 360;
  } else if (wrapped <= -180) {
    wrapped += 360;
  }
  return wrapped;
}

/** Signed shortest rotation from `start` to `end` in (-180, 180]. */
export function shortestDeltaDeg(start: number, end: number): number {
  return normalizeDeg(end - start);
}

/** Interpolate along the shortest arc; 359 to 1 passes through 0. */
export function lerpAngleDeg(start: number, end: number, t: number): number {
  return normalizeDeg(start + shortestDeltaDeg(start, end) * t);
}

export function clamp(value: number, minimum: number, maximum: number): number {
  if (minimum > maximum) {
    throw new Error(`clamp range is inverted: [${minimum}, ${maximum}]`);
  }
  return Math.max(minimum, Math.min(maximum, value));
}

File: backend/app/domain/math2d/angles.py
"""Angle helpers. Serialized angles are counterclockwise degrees."""

from __future__ import annotations

import math


def deg_to_rad(degrees: float) -> float:
    return math.radians(degrees)


def rad_to_deg(radians: float) -> float:
    return math.degrees(radians)


def normalize_deg(degrees: float) -> float:
    """Normalize to the half-open interval (-180, 180]."""
    wrapped = math.fmod(degrees, 360.0)
    if wrapped > 180.0:
        wrapped -= 360.0
    elif wrapped <= -180.0:
        wrapped += 360.0
    return wrapped


def shortest_delta_deg(start: float, end: float) -> float:
    """Signed shortest rotation from ``start`` to ``end`` in (-180, 180]."""
    return normalize_deg(end - start)


def lerp_angle_deg(start: float, end: float, t: float) -> float:
    """Interpolate along the shortest arc; 359 to 1 passes through 0."""
    return normalize_deg(start + shortest_delta_deg(start, end) * t)


def clamp(value: float, minimum: float, maximum: float) -> float:
    if minimum > maximum:
        raise ValueError(f"clamp range is inverted: [{minimum}, {maximum}]")
    return max(minimum, min(maximum, value))
